File: maximum-subarray-min-product.cc

#define FOR(i, a, b) for (remove_cv<remove_reference<decltype(b)>::type>::type i = (a); i < (b); i++)
#define REP(i, n) FOR(i, 0, n)
#define ROF(i, a, b) for (remove_cv<remove_reference<decltype(b)>::type>::type i = (b); --i >= (a); )

class Solution {
public:
  int maxSumMinProduct(vector<int>& a) {
    long n = a.size(), ans = 0;
    vector<int> l(n), r(n);
    vector<long> s(n+1);
    REP(i, n) {
      int x = i;
      while (x && a[x-1] >= a[i])
        x = l[x-1];
      l[i] = x;
      s[i+1] = s[i]+a[i];

File: maximum-subarray-min-product.rs
impl Solution {
    pub fn max_sum_min_product(a: Vec<i32>) -> i32 {
        let mut st = vec![];
        let mut l: Vec<usize> = vec![0; a.len()];
        let mut r: Vec<usize> = vec![a.len(); a.len()];
        for (i, &x) in a.iter().enumerate() {
            while let Some(&j) = st.last() {
                if a[j] < x { break; }
                r[j] = i;
                st.pop();
            }
            if let Some(&j) = st.last() { l[i] = j+1; }
            st.push(i);
        }
        let sum: Box<[i64]> = std::iter::once(0).chain(a.iter().scan(0, |s,&y| { *s += y as i64; Some(*s) })).collect();
        let mut ans = 0;
        for i in 0..a.len() {
            ans = ans.max((sum[r[i]]-sum[l[i]]) * a[i] as i64);
        }
        (ans % 1000000007) as i32
    }
}

///

impl Solution {
    pub fn max_sum_min_product(a: Vec<i32>) -> i32 {
        use std::{iter,usize::MAX};
        let sum: Box<[i64]> = iter::once(0).chain(a.iter().scan(0, |s,&y| { *s += y as i64; Some(*s) })).collect();
        